### ISSUE TITLE
Preserve callstack

### ISSUE THREAD

**Issue by rkjdid** [aio-libs/async-lru#528]
After using the lib I realized callstack was not preserved, which is a (minor) issue for my logging system - I put a marker higher up the stack for logging library to set stacklevel to a higher level frame, while lru_cache is set on low level functions.

I understand why and how this happens with the new task pushed in event_loop, but please bear in mind I'm only starting to switch to a system including async parts, so my understanding is limited.

I implemented a radical workaround, dismissing the future + callback mechanism here: https://github.com/aio-libs/async-lru/commit/8bf9eeadbbc132f9a494dab8689dd5774d3001b8  - two features seem to be crippled: ttl does not seem to be possible, and also consecutive calls while first one is not done yet will always miss cache (since cache no longer stores future but result only after it's available)

2 questions :

  - are there any other side effect to this patch?
  - would it be possible to do this properly?

**Comment by Dreamsorcerer** [comment]
Your code doesn't handle multiple calls properly, which is the point of all the code you removed.

i.e.
```
await asyncio.gather(*(my_func(5) for _ in range(5))
```

Results in running the body of your function 5 times. Our code would result in it running once.

**Comment by Dreamsorcerer** [comment]
I think you already reached that conclusion anyway, but that's the reason for the code. I wouldn't know anything more, unless any test failures give you a hint.

**Comment by Dreamsorcerer** [comment]
I think a task can often tell you where it's created, but a full stacktrace is probably not going to happen. Unless you grab the stacktrace when you call the function and pass it through as somewhere, which the logger can pick up, or something awkward like that...

**Comment by Dreamsorcerer** [comment]
Although, actually, you should get a full traceback when `await`ing...
e.g.
```
import async_lru, asyncio

def raise_something():
    raise Exception()

@async_lru.alru_cache()
async def foo():
    raise_something()

async def main():
    await foo()

asyncio.run(main())
```
Produces:
```
Traceback (most recent call last):
  File "/home/ubuntu/desktop/test.py", line 14, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/home/ubuntu/desktop/test.py", line 11, in main
    await foo()
  File "/home/ubuntu/.local/lib/python3.10/site-packages/async_lru/__init__.py", line 227, in __call__
    return await asyncio.shield(fut)
  File "/home/ubuntu/desktop/test.py", line 8, in foo
    raise_something()
  File "/home/ubuntu/desktop/test.py", line 4, in raise_something
    raise Exception()
Exception
```

So, what's the actual call trace you get?

**Comment by rkjdid** [comment]
But I agree that it seems somehow when raising error it does seem to recollect all the stack..

**Comment by rkjdid** [comment]
leaving this here to illustrate difference between printing stack while in the func vs after task is done (after raising)

```
  File "/home/rkj/src/octd/oop.py", line 16, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
  File "/usr/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
  File "/usr/lib/python3.11/asyncio/base_events.py", line 640, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.11/asyncio/base_events.py", line 607, in run_forever
    self._run_once()
  File "/usr/lib/python3.11/asyncio/base_events.py", line 1922, in _run_once
    handle._run()
  File "/usr/lib/python3.11/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/home/rkj/src/octd/oop.py", line 6, in bar
    traceback.print_stack()


== running stack vs. raise stack ==


Traceback (most recent call last):
  File "/home/rkj/src/octd/oop.py", line 16, in <module>
    asyncio.run(main())
  File "/usr/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/rkj/src/octd/oop.py", line 14, in main
    await foo()
  File "/home/rkj/src/octd/oop.py", line 11, in foo
    await bar()
  File "/home/rkj/v/octd/lib/python3.11/site-packages/async_lru/__init__.py", line 227, in __call__
    return await asyncio.shield(fut)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rkj/src/octd/oop.py", line 8, in bar
    raise Exception()
Exception
```

**Comment by Dreamsorcerer** [comment]
Yep, sorry, been jumping between different tasks. So, when you run a print/log in the task, it will be run in the context of that task, and the task itself is basically the entry point of the code (because the task is created and then scheduled in the loop, so the loop then enters the task). By comparison, when an exception is raised, it is essentially stored in the task, and then when `await`ed later the stored exception is resolved in the context of the `await` call and thus gets put together into the fuller traceback seen above.

The callstack you're looking for simply doesn't exist. What if the task never gets 'await'ed, or is 'await'ed multiple times, then what callstack are you expecting to see?

I think it's possible to get the location that a task was created, although I don't remember how exactly (try looking at the attributes on the task object), but not sure you can get the callstack at the point it was created.

**Comment by rkjdid** [comment]
Yep thanks it all makes sense.

I think the proper (and only) way to achieve this is via `contextvars`, it seems completely suited for this and I manage to achieve what I want using this straightforward approach, not altering current code.  https://github.com/aio-libs/async-lru/commit/66df86260db48404d9e968acac8deda5075f2a5a - note the context is conveniently copied in `create_task` so no issue resetting the value right after task creation, all this is synchronous code. 

Now as to if this is relevant for main repo I don't know: I don't think it has any downsides, but probably few use-cases too.

**Comment by Dreamsorcerer** [comment]
You could presumably do the same thing with an extra decorator?

Vaguely along the lines of:
```
def record_stack(f):
    def inner(*args, **kwargs):
        stack = inspect.stack()
        return f(*args, **kwargs)
    return inner

@record_stack
@alru_cache
def foo(): ...
```

It might miss the call inside of alru_cache, but that's probably not the part you're looking for.

**Comment by rkjdid** [comment]
Yes it will work like this, I will work with my fork though for now since it's basically always the case that I want to merge these two decorators, bloats the code unnecessarily.

Ideally I would do this from the decorator I actually use for marking the func I want to log, but it does not work since I'm not deep enough in the stack at this point.

```
@alru_cache
def cached_req():
  # low-level log, looking for context from marked func
  pass

@mark_for_logging
def high_level_func():
  return cached_req()

def mark_for_logging(fn):
  def inner(*args, **kwargs):
     # stack not deep enough here for recording
     return fn()
   return inner
```

Thanks for the discussion and help :)
